lib/Dialect/Secret/Transforms/DistributeGeneric.cpp

```diff
@@ -0,0 +1,203 @@
+#include "include/Dialect/Secret/Transforms/DistributeGeneric.h"
+
+#include "include/Dialect/Secret/IR/SecretOps.h"
+#include "mlir/include/mlir/IR/PatternMatch.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+namespace secret {
+
+#define GEN_PASS_DEF_SECRETDISTRIBUTEGENERIC
+#include "include/Dialect/Secret/Transforms/Passes.h.inc"
+
+// Inline the inner block of a secret.generic that has no secret operands.
+//
+// E.g.,
+//
+//    %res = secret.generic ins(%value : i32) {
+//     ^bb0(%clear_value: i32):
+//       %c7 = arith.constant 7 : i32
+//       %0 = arith.muli %clear_value, %c7 : i32
+//       secret.yield %0 : i32
+//    } -> (!secret.secret<i32>)
+//
+// is transformed to
+//
+//    %0 = arith.constant 0 : i32
+//    %res = arith.muli %value, %0 : i32
+//
+struct CollapseSecretlessGeneric : public OpRewritePattern<GenericOp> {
+  CollapseSecretlessGeneric(mlir::MLIRContext *context)
+      : OpRewritePattern<GenericOp>(context, /*benefit=*/2) {}
+
+  LogicalResult matchAndRewrite(GenericOp op,
+                                PatternRewriter &rewriter) const override {
+    for (Type ty : op.getOperandTypes()) {
+      if (dyn_cast<SecretType>(ty)) {
+        return failure();
+      }
+    }
+
+    YieldOp yieldOp = dyn_cast<YieldOp>(op.getBody()->getOperations().back());
+    rewriter.inlineBlockBefore(op.getBody(), op.getOperation(), op.getInputs());
+    rewriter.replaceOp(op, yieldOp.getValues());
+    rewriter.eraseOp(yieldOp);
+    return success();
+  }
+};
+
+// Remove unused args of a secret.generic op
+//
+// E.g.,
+//
+//    %res = secret.generic
+//       ins(%value_sec, %unused_sec : !secret.secret<i32>, !secret.secret<i32>)
+//       {
+//     ^bb0(%used: i32, %unused: i32):
+//       %0 = arith.muli %used, %used : i32
+//       secret.yield %0 : i32
+//    } -> (!secret.secret<i32>)
+//
+// is transformed to
+//
+//    %res = secret.generic
+//       ins(%value_sec : !secret.secret<i32>) {
+//     ^bb0(%used: i32):
+//       %0 = arith.muli %used, %used : i32
+//       secret.yield %0 : i32
+//    } -> (!secret.secret<i32>)
+//
+struct RemoveUnusedGenericArgs : public OpRewritePattern<GenericOp> {
+  RemoveUnusedGenericArgs(mlir::MLIRContext *context)
+      : OpRewritePattern<GenericOp>(context, /*benefit=*/1) {}
+
+  LogicalResult matchAndRewrite(GenericOp op,
+                                PatternRewriter &rewriter) const override {
+    bool hasUnusedOps = false;
+    Block *body = op.getBody();
+    for (BlockArgument arg : body->getArguments()) {
+      if (arg.use_empty()) {
+        hasUnusedOps = true;
+        rewriter.updateRootInPlace(op, [&]() {
```

Comment:
Is it safe to remove by index? Would next iteration `arg.getArgNumber()` be decremented "magically" ?

Reply:
It looks like the implementation does indeed update argNumber for all arguments: https://github.com/llvm/llvm-project/blob/9c5b38b49fa2a1a81dac78faa0a8432eff495615/mlir/lib/IR/Block.cpp#L192

Reply:
Update, you were right :) It passes tests, but hits an assertion which apparently only runs in debug mode internally

Updating to

```
    for (int i = 0; i < body->getArguments().size(); ++i) {
      BlockArgument arg = body->getArguments()[i];
      if (arg.use_empty()) { 
        hasUnusedOps = true;  
        rewriter.updateRootInPlace(op, [&]() {
          body->eraseArgument(i);
          op.getOperation()->eraseOperand(i);
        });
        // Ensure the next iteration uses the right arg number
        --i;
      }
    }
```